lib/compress/reduce-vars.js

```diff
@@ -528,93 +539,149 @@ function handle_defined_after_hoist(parent) {
-                if (symbols_of_interest.has(id)) {
-                    if (node instanceof AST_SymbolDeclaration || is_lhs(node, tw)) {
-                        found_symbol_writes.add(found_symbols.length);
-                    }
-                    found_symbols.push(id);
+        }
+        // if we found a defun on the list, mark IN_DEFUN=id and descend
+
+        if (node instanceof AST_Symbol && node.thedef) {
+            const id = node.definition().id;
+            if (symbols_of_interest.has(id)) {
+                if (node instanceof AST_SymbolDeclaration || is_lhs(node, tw)) {
+                    found_symbol_writes.add(found_symbols.length);
                 }
+                found_symbols.push(id);
             }
-        })));
-
-        for (const { def, defun, fname_def } of potential_conflicts) {
-            const defun_range = defun_ranges.get(defun);
-
-            // find the index in `found_symbols`, with some special rules:
-            const find = (sym_id, starting_at = 0, must_be_write = false) => {
-                let index = starting_at;
-
-                for (;;) {
-                    index = found_symbols.indexOf(sym_id, index);
-
-                    if (index === -1) {
-                        break;
-                    } else if (index >= defun_range.start && index < defun_range.end) {
-                        index = defun_range.end;
-                        continue;
-                    } else if (must_be_write && !found_symbol_writes.has(index)) {
-                        index++;
-                        continue;
-                    } else {
-                        break;
-                    }
-                }
+        }
+    })));
+
+    const defun_first_read_map = new Map();
+    for (const defun of defuns_of_interest) {
+        const defun_range = defun_ranges.get(defun);
+        const defun_first_read = found_symbols.findIndex(
+            (sym, index) =>
+                sym === defun &&
+                (index < defun_range.start || index > defun_range.end)
+        );
+        if (defun_first_read === -1) {
+            continue;
+        }
+        defun_first_read_map.set(defun, defun_first_read);
+
+        // Update all depdencies of `defun`
+        let queue = enclosing_defuns_map.get(defun);
+        queue = queue ? queue.slice() : [];
+        const visited = new Set();
+        while (queue.length) {
+            const enclosed_defun = queue.pop();
+            // bail out on circular dependencies
+            if (visited.has(enclosed_defun)) {
+                continue;
+            }
+            visited.add(enclosed_defun);
 
-                return index;
-            };
+            let enclosed_defun_first_read = defun_first_read_map.get(enclosed_defun);
+            if (enclosed_defun_first_read == null) enclosed_defun_first_read = Infinity;
+            if (enclosed_defun_first_read < defun_first_read) {
+                continue;
+            }
```

Comment:
```suggestion
            let enclosed_defun_first_read = defun_first_read_map.get(enclosed_defun);
            if (enclosed_defun_first_read != null && enclosed_defun_first_read < defun_first_read) {
                continue;
            }
```

Seems like we can simplify the if-statement by not using Infinity